### PR TITLE
Implement Refresh Token functionality and tests

### PR DESCRIPTION
**Pull request by Dharakkkk**
This commit includes the implementation of refresh tokens functionality in the auth controller, as well as associated tests. Here are the detailed notes on the changes made:

Implementation of the POST /api/auth/refresh endpoint in the auth controller. This function retrieves the refresh token from the request, validates it, and creates a new access token and refresh token. The new refresh token is then added to the response as a cookie.

Implementation of a series of unit tests for the aforementioned function. These tests cover different scenarios such as the case when both the access token and the refresh token are valid, absence of the refresh token, incorrect access token, incorrect refresh token, and the case when the access token has expired.

Moving the RefreshTokenTest tests to the 'controller' package, since these tests are focused on the behavior of the auth controller.

Updating the names of the tests to be more descriptive and adding DisplayName annotations to each test to make it easier to understand what each test does.

This commit aims to provide refresh tokens functionality, which is essential for the security and convenience of users.

#### Checklist template for LEADERS

&#x2611; | &#x2610;

&#x2610; New code is covered with tests and coverage requirements are met.
&#x2610; Build is not failing.
&#x2610; All tests are passing.
&#x2610; All conflicts are resolved and the branch is rebased on top of the up-to-date parent branch.
&#x2610; All issue requirements are implemented.
&#x2610; At least two reviewers are assigned.
&#x2610; The Pull request is linked to the related issue.
&#x2610; Correct branch name. Convention [https://github.com/bycza-zagroda/track-expenses-app-documentation/blob/develop/githubwork/GITHUB_WORK.md]
&#x2610; Correct unit test name - suffix 'Test' in the end.
&#x2610; Correct integration test name - suffix 'IT' in the end.
&#x2610; Positive and negative test paths have been implemented.

### CODE REVIEW

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/config/WebSecurityConfig.java [hunk 1]
```diff
@@ -33,6 +33,7 @@ public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
                 .authorizeHttpRequests(authorize -> authorize
                         .antMatchers("/").permitAll()
                         .antMatchers("/api/auth/login").permitAll()
+                        .antMatchers("/api/auth/refresh").authenticated()
```
Comment (poneciak57) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/config/WebSecurityConfig.java:36]: user should be able to reach refresh endpoint when he is unauthenticated, thats the whole point of refreshing
users token expired -> user refreshes the token -> user is able to continue using api

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/AuthController.java [hunk 2]
```diff
@@ -45,8 +48,21 @@ public ResponseEntity<Void> registerUser(@Valid @RequestBody AuthRegisterDTO aut
     }
 
     @PostMapping("/refresh")
-    public ResponseEntity<AuthAccessTokenDTO> refreshToken(@CookieValue(name = "refresh_token") String refreshToken) {
-        //TODO
-        return new ResponseEntity<>(new AuthAccessTokenDTO("TODO-accessTOKEN"), HttpStatus.OK);
+    public ResponseEntity<AuthAccessTokenDTO> refreshToken(@CookieValue(name = "refresh_token", required = false) String refreshToken, HttpServletRequest request, HttpServletResponse response) {
```
Comment (poneciak57) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/AuthController.java:51]: refresh token is required

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/AuthController.java [hunk 3]
```diff
@@ -45,8 +48,21 @@ public ResponseEntity<Void> registerUser(@Valid @RequestBody AuthRegisterDTO aut
...
+        String accessToken = request.getHeader("Authorization").replace("Bearer ", "");
+
+        if (userService.validateToken(accessToken) && userService.validateToken(refreshToken)) {
```
Comment (poneciak57) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/AuthController.java:54]: access token should have valid signature but can be expired in that case
on the other hand refresh token should have valid signature as well as cannot be expired

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/AuthController.java [hunk 4]
```diff
@@ -45,8 +48,21 @@ public ResponseEntity<Void> registerUser(@Valid @RequestBody AuthRegisterDTO aut
...
+            User user = userService.getUserFromToken(accessToken);
+
+            String newAccessToken = userService.createAccessToken(user);
+            Cookie newRefreshTokenCookie = userService.createRefreshTokenCookie(user);
+
+            response.addCookie(newRefreshTokenCookie);
+
+            return new ResponseEntity<>(new AuthAccessTokenDTO(newAccessToken), HttpStatus.OK);
+        }
```
Comment (poneciak57) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/controller/AuthController.java:63]: I think the whole logic of this endpoint should be in service

File: src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/RefreshTokenTest.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+        mockMvc.perform(post("/api/auth/refresh")
+                        .cookie(new Cookie("refresh_token", invalidRefreshToken))
+                        .header("Authorization", "Bearer " + validAccessToken))
+                .andExpect(status().isUnauthorized());
+    }
+
+    @DisplayName("Attempt to refresh tokens with expired access token")
+    @Test
+    void refreshTokenWithExpiredAccessToken_thenReturnUnauthorized() throws Exception {
```
Comment (poneciak57) [inline src/test/java/pl/byczazagroda/trackexpensesappbackend/controller/RefreshTokenTest.java:112]: as i mentioned earlier user should be able to refresh expired access token in particular

**Review by poneciak57 — CHANGES_REQUESTED** [review]
Requested changes

**Review by dawciobiel — APPROVED** [review]
LGTM

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImpl.java [hunk 6]
```diff
@@ -131,6 +136,56 @@ private Cookie createRefreshTokenCookie(User user) {
...
+    public String refreshToken(HttpServletRequest request, HttpServletResponse response,
+                               String refreshToken, String accessToken) {
+        if(!validateTokenSignature(refreshToken) && !validateTokenExpiry(refreshToken)) {
+            throw new AppRuntimeException(ErrorCode.U004,
+                    "Refresh token is not valid");
+        }
+
+        String accessTokenWithoutBearer = accessToken.replace("Bearer ", "");
+        if(!validateTokenSignature(accessTokenWithoutBearer)) {
+            throw new AppRuntimeException(ErrorCode.U004,
```
Comment (dawciobiel) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImpl.java:176]: Please create a new error code in `ErrorCodes` - for example:
```java
S003("S003", "REFRESH_TOKEN_NOT_VALID", 403); // 401, 403 or other proper HTTP code.
```
**Links**
https://en.wikipedia.org/wiki/List_of_HTTP_status_codes

File: src/main/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImpl.java [hunk 7]
```diff
@@ -131,6 +136,56 @@ private Cookie createRefreshTokenCookie(User user) {
...
+
+    public String refreshToken(HttpServletRequest request, HttpServletResponse response,
+                               String refreshToken, String accessToken) {
+        if(!validateTokenSignature(refreshToken) && !validateTokenExpiry(refreshToken)) {
+            throw new AppRuntimeException(ErrorCode.U004,
+                    "Refresh token is not valid");
+        }
+
+        String accessTokenWithoutBearer = accessToken.replace("Bearer ", "");
+        if(!validateTokenSignature(accessTokenWithoutBearer)) {
```
Comment (dawciobiel) [inline src/main/java/pl/byczazagroda/trackexpensesappbackend/service/UserServiceImpl.java:175]:
```suggestion
        if (!validateTokenSignature(accessTokenWithoutBearer)) {
```

**Review by poneciak57 — APPROVED** [review]
Looks good to me

**Review by dawciobiel — APPROVED** [review]
LGTM

**Pull request merged**